geth/jail/jail.go

```diff
@@ -4,251 +4,266 @@ import (
-	if err != nil {
-		throwJSException(err)
-	}
+func (j *Jail) cell(chatID string) (*Cell, error) {
+	j.cellsMx.RLock()
+	defer j.cellsMx.RUnlock()
 
-	rpc := jail.nodeManager.RPCClient()
-	// TODO(divan): remove this check as soon as jail cells have
-	// proper cancellation mechanism implemented.
-	if rpc == nil {
-		throwJSException(fmt.Errorf("Error getting RPC client. Node stopped?"))
+	cell, ok := j.cells[chatID]
+	if !ok {
+		return nil, fmt.Errorf("cell '%s' not found", chatID)
 	}
-	response := rpc.CallRaw(request.String())
 
-	// unmarshal response to pass to otto
-	var resp interface{}
-	err = json.Unmarshal([]byte(response), &resp)
+	return cell, nil
+}
+
+// Cell returns a cell by chatID. If it does not exist, error is returned.
+// Required by the Backend.
+func (j *Jail) Cell(chatID string) (common.JailCell, error) {
+	return j.cell(chatID)
+}
+
+// Call executes the `call` function within a cell with chatID.
+// Returns a string being a valid JS code. In case of a successful result,
+// it's {"result": any}. In case of an error: {"error": "some error"}.
+//
+// Call calls commands from `_status_catalog`.
+// commandPath is an array of properties to retrieve a function.
+// For instance:
+//   `["prop1", "prop2"]` is translated to `_status_catalog["prop1"]["prop2"]`.
+func (j *Jail) Call(chatID, commandPath, args string) string {
+	cell, err := j.cell(chatID)
 	if err != nil {
-		throwJSException(fmt.Errorf("Error unmarshalling result: %s", err))
+		return newJailErrorResponse(err)
 	}
-	respValue, err := jail.vm.ToValue(resp)
+
+	value, err := cell.Call("call", nil, commandPath, args)
 	if err != nil {
-		throwJSException(fmt.Errorf("Error converting result to Otto's value: %s", err))
+		return newJailErrorResponse(err)
 	}
 
-	return respValue
+	return newJailResultResponse(value)
 }
 
-func newErrorResponse(msg string, id interface{}) map[string]interface{} {
-	// Bundle the error into a JSON RPC call response
-	return map[string]interface{}{
-		"jsonrpc": "2.0",
-		"id":      id,
-		"error": map[string]interface{}{
-			"code":    -32603, // Internal JSON-RPC Error, see http://www.jsonrpc.org/specification#error_object
-			"message": msg,
-		},
+// GetRPCClient returns an rpc.Client.
+func (j *Jail) GetRPCClient() *rpc.Client {
+	if j.rpcClientProvider == nil {
+		return nil
 	}
-}
 
-func newErrorResponseOtto(vm *vm.VM, msg string, id interface{}) otto.Value {
-	// TODO(tiabc): Handle errors.
-	errResp, _ := json.Marshal(newErrorResponse(msg, id))
-	errRespVal, _ := vm.Run("(" + string(errResp) + ")")
-	return errRespVal
+	return j.rpcClientProvider.RPCClient()
 }
 
-func newResultResponse(vm *otto.Otto, result interface{}) otto.Value {
-	resp, _ := vm.Object(`({"jsonrpc":"2.0"})`)
-	resp.Set("result", result) // nolint: errcheck
+// sendRPCCall executes a raw JSON-RPC request.
+func (j *Jail) sendRPCCall(request string) (interface{}, error) {
+	client := j.GetRPCClient()
+	if client == nil {
+		return nil, ErrNoRPCClient
+	}
```

Comment:
(out of scope) Why can it be the case, by the way? Can we avoid it by construction?

Reply:
It can happen when `sendAsync` handler is called and at the same time, the node is not ready or stopped.

Reply:
And this line is about to be gone when that PR with proper mutex guard is merged, right?

Reply:
If we had a graceful node shutdown process, it would cancel or wait to finish all started calls. So, I would say yes if we lived in a perfect world :) However, if a public method can return `nil`, I think we should still check for that.